MovieHacker/ImageCache.swift

```diff
@@ -0,0 +1,58 @@
+//
+//  ImageCache.swift
+//  MovieHacker
+//
+//  Created by Jan Dědeček on 13/08/2017.
+//  Copyright © 2017 Jan Dědeček. All rights reserved.
+//
+
+import UIKit
+
+private let imageCache = NSCache<NSString, UIImage>()
+
+private var tasks = [URL: (URLSessionDataTask, [(UIImage?) -> Void])]()
+
+func image(from url: URL) -> UIImage? {
+  return imageCache.object(forKey: url.absoluteString as NSString)
+}
+
+func loadImage(from url: URL, completion: @escaping (UIImage?) -> Void) {
+  if let image = imageCache.object(forKey: url.absoluteString as NSString) {
+    completion(image)
+    return
+  }
+
+  if tasks[url] != nil {
+    tasks[url]?.1.append(completion)
+    return
+  }
+
+  print("Loading \(url)")
+
+  let task = URLSession.shared.dataTask(with: url) { (data, response, error) in
```

Comment:
Add comment: "We're fetching data, callback is invoked in some background thread. UI is not thread-safe and can only be manipulated from the main thread. Therefore we will use dispatch queues from GCD to ensure that the critical code run on main thread."